### DISCUSSION TITLE
[12장] Getter의 사용

### DISCUSSION THREAD

**Discussion by firewoody237** [cowaine/is-my-code-that-strange#111] [Chapter 11-12 (주석, 메서드)] [answered]
> 묻지 말고 명령하라
> 어떤 클래스가 다른 클래스의 상태를 판단하거나, 상태에 따라 변경을 하면 응집도가 낮은 구조이다.
> `getter` / `setter`는 일반적으로 개발 생산성이 좋지 않은 소프트웨어에서 자주 볼 수 있다.

책에는 위와 같이 나와있는데요, `setter`의 경우 저도 메서드로 풀어내고 실제로 사용하진 않습니다.
근데 `getter`의 경우 사용하지 않으면 어색한 경우가 많은데요, 예를들면 값 객체에서

```
class Name {
    private final String value;
    ....
}
```

위와 같은경우 이름을 받기 위해선 `getName`말고는 딱히 더 나은 메서드명이 떠오르지 않는데요,
여러분들이라면 그냥 getter를 쓰시겠나요, 아니면 메서드로 풀어 사용하실건가요?
(개발하다보면 위와 같은 값 객체 외에도 `getter`가 들어가야 할 것 같은 유혹 포인트가 매우 많았습니다.)

**Comment by Irisation23** [comment, marked as answer]
- 팀에서 getter를 사용하고 있다면 getter를 이용할 것 같습니다.
- 상황에 따라 달라질 것 같습니다.

1. 단순하게 객체의 값을 이용할 때 -> getter 사용
2. 민감한 데이터의 캡슐화를 깰 수 있는 경우 -> getter 이외의 메서드명 사용

- 민감한 데이터의 캡슐화를 깰 수 있는 경우 예제

``` java
public class User {
    private String username;
    private String password;  // 비밀번호에 @Getter 어노테이션 사용하지 않음

    public User(String username, String password) {
        this.username = username;
        setPassword(password);
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public void setPassword(String password) {
        // 비밀번호를 암호화하여 저장
        this.password = encryptPassword(password);
    }

    private String encryptPassword(String password) {
        // 간단한 암호화 로직 (예를 들어 실제 환경에서는 더 복잡한 로직을 사용)
        return "encrypted-" + password;
    }

    // 암호화된 비밀번호는 외부에 노출하지 않음
}
```

> 따라서 상황에 맞는 코드를 작성할 것 같습니다. 개인적으로는 나무꾼 이론이 너무 와닿기 때문에 getter 이용이 좀 더 땡깁니다... (마음은 아프지만)

**Comment by corock** [comment]
저는 DTO 나 JavaBeans 스펙 준수가 필요한 프레임워크에서는 `getXxx` 로 사용하고, 이외의 경우에는 메서드로 푸는 것을 선호합니다.

예시 코드인 `Name` 클래스의 경우 필드나 메서드명 이름 짓기가 까다로워 보이긴 하네요;
만약 성과 이름을 구분해야 하는 등의 요건이 있다면 그나마 명명이 좀 수월해질 것 같긴 합니다.

```java
class Name {

    private final String firstName;
    private final String lastName;

    public String getFullName() {
        return firstName + " " + lastName;
    }

}
```
